PMax Assets: Integrate WP Media library to implement the asset images selector

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implements a part of 📌 [Assets form and related components](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-misc) in #1787.

- Multiple images selector

### Screenshots:

#### 📹 Crop, skip cropping, check image size, remove selected image, and replace selected image

https://user-images.githubusercontent.com/17420811/208123491-7a403bb0-17e8-41a5-8edf-62f01b39e9bd.mp4

#### 📹 Pre-select when replacing image and check the size of the newly uploaded image

https://user-images.githubusercontent.com/17420811/208124129-29c80288-314d-421e-a271-a42ea5d2ddb9.mp4

#### 📹 Remove the deleted image from the selected images

https://user-images.githubusercontent.com/17420811/208124578-711f1ef8-2772-4412-aa3e-66f4f167d29b.mp4

#### 📹 Swap images if the selected image already exists while replacing

https://user-images.githubusercontent.com/17420811/208124807-a5ca07cf-ccf8-447c-b4ed-a5efbd5a03c7.mp4

#### 📹 The image to be replaced has been deleted

https://user-images.githubusercontent.com/17420811/208124844-b7c72ac1-bedb-4e78-810f-163f7ad83438.mp4

#### 📷 Disable "Add image" button and multiline images

![image](https://user-images.githubusercontent.com/17420811/208123458-17c4fe4e-229b-4092-8f07-8353a4203fd1.png)

### Detailed test instructions:

1. Go to the assets step of the campaign creation page.
2. Use the asset image selector to see if most of the operations are working properly.

### Changelog entry

eason9487 committed:
Add a new hook to select cropped images via WP Media library

eason9487 committed:
Implement the ImagesSelector component for select asset images

eason9487 committed:
Add the temporary demo for ImagesSelector component

eason9487 committed:
Avoid not having the `id` when converting `initialImageUrls` prop to the images state in ImagesSelector

eason9487 committed:
Use another way to calculate the percent error of aspect ratio in `useCroppedImageSelector` hook

puntope commented on js/src/hooks/types.js:
```diff
@@ -0,0 +1,12 @@
+/**
```
💅  Since is there not a consensus yet about this `types.js` standard and `ImageMedia` is only being used in `useCroppedImageSelector.js` I would add it there directly deleting this file.

eason9487 replied:
The reason is similar to https://github.com/woocommerce/google-listings-and-ads/pull/1805#discussion_r1044303992. I would like to advocate that the type definitions of hooks be centralized in this file.

puntope replied:
Yes. I understand the idea. But I would suggest to have a team consensus. Maybe a P2 or an agenda question to be all in the same page about the new proposal.

eason9487 replied:
Surely it's nicer to have a team consensus for this, and the discussion of this convention perhaps could be covered in #195.

puntope commented on js/src/hooks/useCroppedImageSelector/useCroppedImageSelector.js:
```diff
@@ -0,0 +1,289 @@
+ */
+
+/**
+ * Calculates the maximum cropping size according to the fixed aspect ratio.
+ *
+ * @param {number} width The image width.
+ * @param {number} height The image height.
+ * @param {number} widthScale The scale of width.
+ * @param {number} heightScale The scale of height.
+ * @return {[number, number]} The tuple of cropped width and height.
+ */
+export function cropByFixedRatio( width, height, widthScale, heightScale ) {
+	const ratio = widthScale / heightScale;
+
+	if ( width / height > ratio ) {
+		width = Math.round( height * ratio );
+	} else {
+		height = Math.round( width / ratio );
+	}
+
+	return [ width, height ];
+}
```
❓  I guess this just calculate the maximum cropping size but is not actually cropping the image.

💅  If so, I would rename the function as  `calculateMaxCropByFixedRatio` or something similar. IMO without reading the function internals `cropByFixedRatio` could be understood as we are actually cropping the image.

eason9487 replied:
Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/d11fc0ec9a053e95a77aadb8dcfd6fca1f76bea3.

puntope commented on js/src/hooks/useCroppedImageSelector/useCroppedImageSelector.js:
```diff
@@ -0,0 +1,289 @@
+				const args = [ width, height, minWidth, minHeight ];
+				const ratioPercentError = calcRatioPercentError( ...args );
+
+				if ( ratioPercentError < allowedRatioPercentError ) {
```
💅  I guess this should be `<= `since we "allowed" that percent error, so equals is valid. Otherwise we could rename it to `ratioPercentErrorTolerance` or just `ratioPercentError`

eason9487 replied:
Thanks for the suggestion. Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/4f321138ba65ec29d6246d4f68d9352edde72702.

puntope commented on js/src/hooks/useCroppedImageSelector/useCroppedImageSelector.js:
```diff
@@ -0,0 +1,289 @@
+
+	return [ width, height ];
+}
+
```
💅  Even tho you added the refs of the extended library. I missed some explanations on each file function, just explaining what we are doing there is enough, no need to be super detailed. I guess tat would save time of checking parent files for the developers landing on this code.

eason9487 replied:
Since the WP Media library seems to have no documents and it's mainly built on top of Backbone.js and mixed some WP custom mechanisms,  the need for additional explanations very depends on the readers' understanding of this background knowledge. Could you help point out which places need more explanation?

puntope replied:
>  Could you help point out which places need more explanation?

Especially I had doubts about these 2. (the second one in L171, for sure)


- In the chunk of code for handling the invalid size and showing error message: (Line 216)

maybe something like; 
// Check if the selected image has a valid width and height. If it has invalid date show an error message in the toolbar. 


- Line 171, about the `states`. Maybe explain why we are loading here this controllers. 

// Maybe like override the WP Media adding suggested width and height for the image and a custom image cropper. 

(Unsure if its accurate)  

Optionally to document the handlers and other internal functions.

eason9487 replied:
Not sure line numbers 171 and 216 are referred to which code revision. Based on the comment content, I guess it means:
- https://github.com/woocommerce/google-listings-and-ads/blob/54da6a7d95648b4b0f8f7f8576960699fc1420df/js/src/hooks/useCroppedImageSelector/useCroppedImageSelector.js#L233
- https://github.com/woocommerce/google-listings-and-ads/blob/54da6a7d95648b4b0f8f7f8576960699fc1420df/js/src/hooks/useCroppedImageSelector/useCroppedImageSelector.js#L198

I added some code comments and adjusted some naming in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/bd25f054a082d990ea94d76759ceac193bd52ef9. Most internal functions still don't have additional documents because the function names should be sufficient to describe themselves. Hope it has covered where the explanations were needed.

puntope commented on js/src/components/paid-ads/asset-group/images-selector.js:
```diff
@@ -0,0 +1,126 @@
+					nextImages.splice( index, 1, { ...replacingImage } );
+				}
+				// If index gets -1 here, it means the image to be replaced has been deleted.
+				index = nextImages.indexOf( replacingImage );
```
❓  I guess this means the we clicked on "Delete permanently" in the media selector?

eason9487 replied:
Yes, but ImagesSelector should be agnostic to the WP Media library, which relies only on the interface of useCroppedImageSelector. I adjusted the code comment a little bit in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/0932206b4476c5bce5080f0798f897af380243a9. Does this make it clearer?

puntope replied:
> Yes, but ImagesSelector should be agnostic to the WP Media library,

Could you elaborate this? Does it mean we can use another media Library rather the WP Media like a 3rd party one?

puntope replied:
❓  If so, should we test the PR against one 3rd party media library to see if it works?

eason9487 replied:
> > Yes, but ImagesSelector should be agnostic to the WP Media library,
> 
> Could you elaborate this? Does it mean we can use another media Library rather the WP Media like a 3rd party one?

It's simply due to the management/practice of interdependence.

Since the JS part of WP Media library is based on jQuery, Backbone.js and some `wp.*` libraries, and it also has its own lifecycle and state management, one of the purposes is to isolate the dependency on WP Media in `useCroppedImageSelector`.

`ImagesSelector`, as a hook user of `useCroppedImageSelector`, relies only on the interfaces provided by the hook for both code implementation and comments, and does not need to be coupled with the WP Media library. For example, `ImagesSelector` doesn't need to know how an image would be deleted, or by whom, but it only needs to know when and which image was deleted.

Therefore, that's the thought of "ImagesSelector should be agnostic to the WP Media library".

The possibility of replacing the implementation of `useCroppedImageSelector` with some other library is the result of this design approach, but not the purpose. In this case, I didn't expect WP Media would be replaced by a 3rd-party library, because in practice this probably won't happen.

eason9487 replied:
BTW, I adjusted the code comment in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/4b111e86bdd3c1869b50316094daae730d7e8273 since the -1 `index` case originally described no longer occurs after the fix in adb4734264ea9ac39e5c87c11b0229c6952757e0.

puntope commented on js/src/components/paid-ads/asset-group/images-selector.js:
```diff
@@ -0,0 +1,126 @@
+			// Find if there is a duplicate image first.
+			let index = nextImages.findIndex( ( { id } ) => id === image.id );
+
+			if ( replacingImage ) {
```
💅  I found `replacingImage` a bit confusing in order to identify if `replacingImage` was

A:  the new selected image to replace the previous one. 
B:  the previous image to be replaced by the new selected image. 

After inspecting all the code I guess its B. So I'd suggest replacing `image` with `newSelectedImage` and `previousSelectedImage`

eason9487 replied:
Consider that the `replacingImage` may not only be replaced but may also be swapped places or removed without further processing. I renamed it to `awaitingActionImage` in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/84f63e919ad5d4e5a12d80475f3c73c560785beb.

puntope replied:
Thanks for the adjustment. I'm still not sure about the new name `awaitingActionImage`. The word `await` and `action` in the name could have ambiguous meanings and I found it confusing. 

I don't want to sound pushy about my suggestion. But what is against `newSelectedImage` and `previousSelectedImage`?  I think it clearly defines what the user is doing in the UI.

eason9487 replied:
> I'm still not sure about the new name `awaitingActionImage`. The word `await` and `action` in the name could have ambiguous meanings and I found it confusing.

It's precisely to avoid ambiguity, the word "selected" is used only in the `onSelect` callback and refers to "the selected image", and the word was avoided in the rest of `ImagesSelector`.

Otherwise, for example, would the `previousSelectedImage` literally mean the previously selected image in the `onSelect` callback, or the previously selected image in the callback of the replacement buttons?

The `awaitingActionImage` keeps the last image data sent from the associated click handler of a replacement button. After "the selected image" is called back from `onSelect`, the next action for `awaitingActionImage` could be:
- pointing out which image to be replaced with the selected image.
- pointing out which image to be swapped the position with the selected image.
- nothing if the reset action for `awaitingActionImage` has occurred in the `onDelete` callback.

Thus, the `awaitingActionImage` is an image **awaiting** something to happen, and when it does, it takes **action** according to the situation at the moment.

---

Besides, I didn't rename `image` to `newSelectedImage` because the `image` as the parameter to `onSelect` callback should be enough to indicate that it is the "selected image", and the `onSelect` implementation is only a few lines long, so I thought it's probably not confusing. But it's just my opinion. I'm happy to rename `image` to `selectedImage` if it's still confusing you.

puntope commented:
Howdy @eason9487 . I tested this PR, thanks for all the code and work here:

I did this checks besides my suggestion in code:

✅  A red message appears near the button when the image size are below the minimum. And it disables the selection.
✅  Images are cropped correctly preserving the desired aspect ratio and the min height and width. 
✅  When image is correctly sized, it allows to skip crop.
✅  The maximum number of image limitation works.
✅  Swap images works (see conflict with remove image at the bottom)
✅  Remove image works
✅  Replace image works
✅  Replace image and then delete the image works.  (see conflict with swap image at the bottom)
✅  Squared images behaviour (works good as previous checked steps)
✅  Pre-select when replacing image and check the size of the newly uploaded image works
✅  Multiline images

❓ **Topic 1**: Set initial crop to suggested size if possible. 

I somehow expected this behaviour, not sure if thats planned or it could be omitted:

- Image squared: Suggested size 1200x1200 
- Image A squared: Original size 1500x1500
- Select image A, open cropper with 1200x1200 preselected.

Current behaviour: 

- Image squared: Suggested size 1200x1200 
- Image A squared: Original size 1500x1500
- Select image A, open cropper with 1500x1500 preselected.

❓ **Topic 2**: Disable crop button if no crop is done. 

 I somehow expected this behaviour, not sure if thats planned or it could be omitted:

- Image squared: Suggested size 1200x1200 
- Image A squared: Original size 1200x1200
- Select image A, open cropper with 1200x1200 preselected.
- No crop was done, so Crop Button is disabled.

Current behaviour: 

- Image squared: Suggested size 1200x1200 
- Image A squared: Original size 1200x1200
- Select image A, open cropper with 1200x1200 preselected.
- No crop was done, but "Crop" button is enabled
- Click on "Crop"
- New cropped image with same dimensions is created. 

**⚠️  The image to be replaced has been deleted in combination with swap images doesn't work as expected:**

I expected this behaviour:

- Add Image A 
- Add Image B
- Click on Image B 
- Delete Image B permanently
- Select image A
- Save with Skip Crop
- Image B is deleted. Image A is in the image list

Current behaviour: 

- Add Image A 
- Add Image B
- Click on Image B 
- Delete Image B permanently
- Select image A
- Save with Skip Crop
- Image B and Image A are in the image list, when clicking on Image B again, there is no selection.


https://user-images.githubusercontent.com/5908855/208607953-e42f2bbb-cc2c-4feb-91b8-139f779c5ff2.mov

eason9487 committed:
Rename the `cropByFixedRatio` function in useCroppedImageSelector to `calcMaxCroppingByFixedRatio`.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1819#discussion_r1052947428

eason9487 committed:
Rename the `allowedRatioPercentError` parameter of useCroppedImageSelector to `ratioPercentError`.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1819#discussion_r1052949168

eason9487 committed:
Update a code comment in ImagesSelector to make it clearer.

https://github.com/woocommerce/google-listings-and-ads/pull/1819#discussion_r1052966097

eason9487 committed:
Rename the `replacingImage` state in ImagesSelector to `awaitingActionImage`.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1819#discussion_r1052967626

eason9487 committed:
Fix the bug that when an image to be replaced has been removed, the removed picture reappears if it's also processed a swapping.

Address the bug in https://github.com/woocommerce/google-listings-and-ads/pull/1819#issuecomment-1358944593

eason9487 committed:
Disable the crop button if no cropping is needed for useCroppedImageSelector.

Address the Topic 2 in https://github.com/woocommerce/google-listings-and-ads/pull/1819#issuecomment-1358944593

eason9487 commented:
@puntope, thanks for the detailed code review and testing.

With this comment, I believe all review comments were addressed or replied. Could you help with a new round of code reviews?

---

> ❓ **Topic 1**: Set initial crop to suggested size if possible. [...]

Since the size and content of the original images have usually been adjusted by users, using the recommended size to pre-select the cropping might annoying to users.

For example, if the user already has many 1500x1500 images and the appropriate gaps between the content and borders have been retained within images. The pre-selected 1200x1200 area in the middle of images may instead always crop onto the content of images, causing the user to keep reselecting to the largest area.

Of course, the opposite case is also true, and there are advantages and disadvantages to either approach, so I think we could go with the current implementation for now.

---

> ❓ **Topic 2**: Disable crop button if no crop is done. [...]

This is a great suggestion! I added this processing by https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/54da6a7d95648b4b0f8f7f8576960699fc1420df.

https://user-images.githubusercontent.com/17420811/209540786-e73f3a84-0754-4907-904a-6e023bacddd3.mp4

---

>  **⚠️  The image to be replaced has been deleted in combination with swap images doesn't work as expected:** [...]

Thanks for catching this. 👍 

Fixed by https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/adb4734264ea9ac39e5c87c11b0229c6952757e0.

https://user-images.githubusercontent.com/17420811/209540876-aa0b52bf-6914-472d-90f9-671711507697.mp4

puntope commented:
Howdy @eason9487 

I tested again the behaviour and now is looking good. Thanks for the adjustments.

Looks good to me. My only concern besides my 💅  comments is this one https://github.com/woocommerce/google-listings-and-ads/pull/1819#discussion_r1057182329

Just to be sure if it would be convenient to test with a 3rd party media lib or not before approval.

eason9487 committed:
Add more code comments for useCroppedImageSelector.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1819#discussion_r1052952059

eason9487 committed:
Update a code comment in ImagesSelector since the case originally described no longer occurs after the fix in adb4734264ea9ac39e5c87c11b0229c6952757e0.

Ref: https://github.com/woocommerce/google-listings-and-ads/commit/adb4734264ea9ac39e5c87c11b0229c6952757e0

eason9487 commented:
Hi @puntope, thanks for the comments. I pushed two commits for adjusting code comments and replied the comments in the second round of code reviews. Could you take another look at this PR?

puntope reviewed (approved):
✅  Approved thanks @eason9487
